### ISSUE TITLE
Queries GitHub API even if there is a binary in the cache

### ISSUE THREAD

**Issue by smessmer** [SergeyPirogov/webdriver_manager#442]
When I run an application with webdriver-manager for a while, and it continuously destroys and creates GeckoDriver instances over time, it starts failing at some point with the GitHub rate limit, even though the first few runs should have downloaded the binary and my application would be totally fine with continuing to use that binary from the cache directory.

My best guess at why this happens is that webdriver-manager queries GitHub for each created GeckoDriver instance to get the current firefox version, and only afterwards checks the cache. And if GitHub returns an error, then webdriver-manager exits saying it couldn't find the binary in the cache even though it was right there.

Proposal for improvement: Don't check the version with GitHub while a binary is in the cache that is still valid. Only check the GitHub version if there is nothing in the cache.

**Comment by JelleSmet-TomTom** [comment]
Confirmed from my side.

**Comment by SergeyPirogov** [comment]
Well, need to provide a fix. Feel free to contribute

**Comment by sergioteula** [comment]
I am also facing this error and it is quite annoying. I will try to help.

**Comment by sergioteula** [comment]
I have just open a new PR to solve this issue: https://github.com/SergeyPirogov/webdriver_manager/pull/458

**Comment by mere-human** [comment]
FYI: This problem can be circumvented if a specific version of a driver is passed instead of 'latest'. The driver version could be obtained beforehand from cache using `DriverCache`.

**Comment by eyaler** [comment]
@aleksandr-kotlyar i just tried with 4.0.0 and fails:

` for i in range(100): GeckoDriverManager(cache_manager=DriverCacheManager(valid_range=1)).install()
...
'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Python310\lib\site-packages\webdriver_manager\firefox.py", line 39, in install
    driver_path = self._get_driver_binary_path(self.driver)
  File "C:\Python310\lib\site-packages\webdriver_manager\core\manager.py", line 40, in _get_driver_binary_path
    file = self._download_manager.download_file(driver.get_driver_download_url(os_type))
  File "C:\Python310\lib\site-packages\webdriver_manager\drivers\firefox.py", line 39, in get_driver_download_url
    resp = self._http_client.get(
  File "C:\Python310\lib\site-packages\webdriver_manager\core\http.py", line 36, in get
    self.validate_response(resp)
  File "C:\Python310\lib\site-packages\webdriver_manager\core\http.py", line 19, in validate_response
    raise ValueError(
ValueError: response body:
{"message":"API rate limit exceeded for XX.XXX.XXX.XXX. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)","documentation_url":"https://docs.github.com/rest/overview/resources-in-the-rest-api#rate-limiting"}

request url:
https://api.github.com/repos/mozilla/geckodriver/releases/tags/v0.33.0
response headers:
{'Date': 'Tue, 22 Aug 2023 10:12:06 GMT', 'Server': 'Varnish', 'Strict-Transport-Security': 'max-age=31536000; includeSubdomains; preload', 'X-Content-Type-Options': 'nosniff', 'X-Frame-Options': 'deny', 'X-XSS-Protection': '1; mode=block', 'Content-Security-Policy': "default-src 'none'; style-src 'unsafe-inline'", 'Access-Control-Allow-Origin': '*', 'Access-Control-Expose-Headers': 'ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-RateLimit-Used, X-RateLimit-Resource, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, Deprecation, Sunset', 'Content-Type': 'application/json; charset=utf-8', 'Referrer-Policy': 'origin-when-cross-origin, strict-origin-when-cross-origin', 'X-GitHub-Media-Type': 'github.v3; format=json', 'X-RateLimit-Limit': '60', 'X-RateLimit-Remaining': '0', 'X-RateLimit-Reset': '1692700867', 'X-RateLimit-Resource': 'core', 'X-RateLimit-Used': '60', 'Content-Length': '280', 'X-GitHub-Request-Id': 'XXXXXXXXXXXXXXXXXXXX'}
`

**Comment by aleksandr-kotlyar** [comment]
> @aleksandr-kotlyar i just tried with 4.0.0 and fails:
> <details><summary></summary>
> 
> 
> ` for i in range(100): GeckoDriverManager(cache_manager=DriverCacheManager(valid_range=1)).install()
> 
> ...
> 
> 'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
> 
> 'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
> 
> 'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
> 
> 'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
> 
> 'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
> 
> 'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
> 
> 'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
> 
> 'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
> 
> 'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
> 
> 'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
> 
> 'C:\\Users\\eyalg\\.wdm\\drivers\\geckodriver\\win64\\v0.33.0\\geckodriver.exe'
> 
> Traceback (most recent call last):
> 
>   File "<stdin>", line 1, in <module>
> 
>   File "C:\Python310\lib\site-packages\webdriver_manager\firefox.py", line 39, in install
> 
>     driver_path = self._get_driver_binary_path(self.driver)
> 
>   File "C:\Python310\lib\site-packages\webdriver_manager\core\manager.py", line 40, in _get_driver_binary_path
> 
>     file = self._download_manager.download_file(driver.get_driver_download_url(os_type))
> 
>   File "C:\Python310\lib\site-packages\webdriver_manager\drivers\firefox.py", line 39, in get_driver_download_url
> 
>     resp = self._http_client.get(
> 
>   File "C:\Python310\lib\site-packages\webdriver_manager\core\http.py", line 36, in get
> 
>     self.validate_response(resp)
> 
>   File "C:\Python310\lib\site-packages\webdriver_manager\core\http.py", line 19, in validate_response
> 
>     raise ValueError(
> 
> ValueError: response body:
> 
> {"message":"API rate limit exceeded for XX.XXX.XXX.XXX. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)","documentation_url":"https://docs.github.com/rest/overview/resources-in-the-rest-api#rate-limiting"}
> 
> 
> 
> request url:
> 
> https://api.github.com/repos/mozilla/geckodriver/releases/tags/v0.33.0
> 
> response headers:
> 
> {'Date': 'Tue, 22 Aug 2023 10:12:06 GMT', 'Server': 'Varnish', 'Strict-Transport-Security': 'max-age=31536000; includeSubdomains; preload', 'X-Content-Type-Options': 'nosniff', 'X-Frame-Options': 'deny', 'X-XSS-Protection': '1; mode=block', 'Content-Security-Policy': "default-src 'none'; style-src 'unsafe-inline'", 'Access-Control-Allow-Origin': '*', 'Access-Control-Expose-Headers': 'ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-RateLimit-Used, X-RateLimit-Resource, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, Deprecation, Sunset', 'Content-Type': 'application/json; charset=utf-8', 'Referrer-Policy': 'origin-when-cross-origin, strict-origin-when-cross-origin', 'X-GitHub-Media-Type': 'github.v3; format=json', 'X-RateLimit-Limit': '60', 'X-RateLimit-Remaining': '0', 'X-RateLimit-Reset': '1692700867', 'X-RateLimit-Resource': 'core', 'X-RateLimit-Used': '60', 'Content-Length': '280', 'X-GitHub-Request-Id': 'XXXXXXXXXXXXXXXXXXXX'}
> 
> `
> </details>

Sorry, I haven't looked into the code of 4.0.0 version yet. In 3.8.6 the cache thing worked fine (i can guarantee). It should not go online if you have valid cache date. May be it has been broken with refactoring between 3.8.6 and 4.0.0. 

Also, i see from your logs that you don't have a github token env variable, and have exceed GitHub-API-Rate-Limit with 60 queries. Set the token and limit will up to 1000 queries per hour for repo in github-actions or 5000 for user.

**Comment by eyaler** [comment]
@aleksandr-kotlyar 

i can confirm its fixed in 3.8.6 and broken in 4.0.0. this is how i tested:
```
from webdriver_manager.firefox import GeckoDriverManager
for i in range(100): GeckoDriverManager().install()
```
should we reopen this bug? or maybe its related to https://github.com/SergeyPirogov/webdriver_manager/issues/596?

**Comment by aleksandr-kotlyar** [comment]
> @aleksandr-kotlyar 
> 
> 
> 
> i can confirm its fixed in 3.8.6 and broken in 4.0.0. this is how i tested:
> 
> ```
> 
> from webdriver_manager.firefox import GeckoDriverManager
> 
> for i in range(100): GeckoDriverManager().install()
> 
> ```
> 
> should we reopen this bug? or maybe its related to https://github.com/SergeyPirogov/webdriver_manager/issues/596?

@eyaler thank you for testing and confirming.
Yeah, I think its #596 now and I suggest the root cause is the same as here. No need to reopen this one.
